File: src/inputs/plugins/home_assistant.py
import asyncio
import logging
import time
from typing import Optional

import aiohttp
from pydantic import Field

from inputs.base import Message, SensorConfig
from inputs.base.loop import FuserInput
from providers.io_provider import IOProvider


class HomeAssistantInputConfig(SensorConfig):
    """
    Configuration for Home Assistant State Input.

    Parameters
    ----------
    base_url : str
        Home Assistant base URL (e.g. http://homeassistant.local:8123).
    token : str
        Long-lived access token from Home Assistant profile.
    entity_ids : str
        Comma-separated list of entity IDs to monitor
        (e.g. light.living_room,switch.fan,climate.bedroom).
    poll_interval : float
        Seconds between state updates (default: 30).
    """

    base_url: str = Field(default="", description="Home Assistant base URL")
    token: str = Field(default="", description="Long-lived access token")
    entity_ids: str = Field(
        default="", description="Comma-separated entity IDs to monitor"
    )
    poll_interval: float = Field(
        default=30.0, description="Seconds between state polls"
    )


class HomeAssistantStateInput(FuserInput[HomeAssistantInputConfig, Optional[list]]):
    """
    Home Assistant state input that polls device states and reports changes.

    Monitors smart home device states and provides updates to the LLM
    when states change, enabling context-aware robot responses.
    """

    def __init__(self, config: HomeAssistantInputConfig):
        """
        Initialize the Home Assistant state input.

        Parameters
        ----------
        config : HomeAssistantInputConfig
            Configuration for the state input.
        """
        super().__init__(config)

        self.io_provider = IOProvider()
        self.messages: list[Message] = []
        self.descriptor_for_LLM = "Home Assistant Device States"

        self.base_url = config.base_url.rstrip("/")
        self.token = config.token
        self.entity_ids = [e.strip() for e in config.entity_ids.split(",") if e.strip()]
        self.poll_interval = config.poll_interval

        self._last_poll_time: float = 0
        self._last_states: dict = {}

        if not self.base_url:
            logging.warning("HomeAssistantStateInput: base_url not provided")
        if not self.token:
            logging.warning("HomeAssistantStateInput: token not provided")
        if not self.entity_ids:
            logging.warning("HomeAssistantStateInput: no entity_ids configured")

    def _get_headers(self) -> dict:
        """Build authorization headers."""
        return {
            "Authorization": f"Bearer {self.token}",
            "Content-Type": "application/json",
        }

    async def _fetch_state(self, entity_id: str) -> Optional[dict]:
        """
        Fetch the current state of a single entity from Home Assistant.

        Parameters
        ----------
        entity_id : str
            The entity ID to fetch.

        Returns
        -------
        Optional[dict]
            State data or None on failure.
        """
        if not self.base_url or not self.token:
            return None

        url = f"{self.base_url}/api/states/{entity_id}"

        try:
            timeout = aiohttp.ClientTimeout(total=10)
            async with aiohttp.ClientSession(timeout=timeout) as session:
                async with session.get(url, headers=self._get_headers()) as response:
                    if response.status == 200:
                        return await response.json()
                    else:
                        logging.error(
                            f"HomeAssistantStateInput: error fetching {entity_id} "
                            f"status={response.status}"
                        )
                        return None
        except asyncio.TimeoutError:
            logging.error(f"HomeAssistantStateInput: timeout fetching {entity_id}")
            return None
        except aiohttp.ClientError as e:
            logging.error(
                f"HomeAssistantStateInput: network error fetching {entity_id}: {e}"
            )
            return None
        except Exception as e:
            logging.error(
                f"HomeAssistantStateInput: unexpected error fetching {entity_id}: {e}"
            )
            return None

    async def _poll(self) -> Optional[list]:
        """
        Poll Home Assistant for device states at configured interval.

        Returns
        -------
        Optional[list]
            List of state dicts when poll interval elapsed, None otherwise.
        """
        current_time = time.time()

        if current_time - self._last_poll_time < self.poll_interval:
            await asyncio.sleep(1.0)
            return None

        self._last_poll_time = current_time
        await asyncio.sleep(1.0)

        if not self.entity_ids:
            return None

        states = []
        for entity_id in self.entity_ids:
            state = await self._fetch_state(entity_id)
            if state is not None:
                states.append(state)

        return states if states else None

    def _format_state(self, state: dict) -> str:
        """
        Format a single entity state into human-readable text.

        Parameters
        ----------
        state : dict
            Raw state data from Home Assistant API.

        Returns
        -------
        str
            Human-readable state description.
        """
        entity_id = state.get("entity_id", "unknown")
        current_state = state.get("state", "unknown")
        attributes = state.get("attributes", {})
        friendly_name = attributes.get("friendly_name", entity_id)

        parts = [f"{friendly_name} ({entity_id}) is {current_state}"]

        if "brightness" in attributes and attributes["brightness"] is not None:
            brightness_pct = round(attributes["brightness"] / 255 * 100)
            parts.append(f"brightness {brightness_pct}%")

        if "color_name" in attributes:
            parts.append(f"color {attributes['color_name']}")

        if "temperature" in attributes:
            parts.append(f"temperature {attributes['temperature']}°C")

        if "current_temperature" in attributes:
            parts.append(f"current temperature {attributes['current_temperature']}°C")

        return ", ".join(parts)

    async def _raw_to_text(self, raw_input: Optional[list]) -> Optional[Message]:
        """
        Convert raw state list to human-readable message, only when states change.

        Parameters
        ----------
        raw_input : Optional[list]
            List of state dicts from Home Assistant.

        Returns
        -------
        Optional[Message]
            Formatted message if states changed, None otherwise.
        """
        if raw_input is None:
            return None

        changed = []
        for state in raw_input:
            entity_id = state.get("entity_id", "")
            current_state = state.get("state", "")
            last_known = self._last_states.get(entity_id)

            if last_known != current_state:
                changed.append(state)
                self._last_states[entity_id] = current_state

        if not changed:
            return None

        lines = [self._format_state(s) for s in changed]
        message = "Smart home device updates: " + "; ".join(lines)
        return Message(timestamp=time.time(), message=message)

    async def raw_to_text(self, raw_input: Optional[list]):
        """
        Update message buffer with processed state data.

        Parameters
        ----------
        raw_input : Optional[list]
            Raw state list to process.
        """
        pending_message = await self._raw_to_text(raw_input)

        if pending_message is not None:
            self.messages.append(pending_message)

    def formatted_latest_buffer(self) -> Optional[str]:
        """
        Format and clear the latest buffer contents.

        Returns
        -------
        Optional[str]
            Formatted string for LLM or None if buffer is empty.
        """
        if not self.messages:
            return None

        latest_message = self.messages[-1]

        result = (
            f"\nINPUT: {self.descriptor_for_LLM}\n// START\n"
            f"{latest_message.message}\n// END\n"
        )

        self.io_provider.add_input(
            self.descriptor_for_LLM,
            latest_message.message,
            latest_message.timestamp,
        )
        self.messages = []

        return result

File: plugins/inputs/home_assistant/home_assistant.go
// Package home_assistant provides an input sensor that polls Home Assistant
// entity states and reports changes to the LLM as they occur.
package home_assistant

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"

	"go.uber.org/zap"

	"github.com/openmind/om1/internal/httpclient"
	"github.com/openmind/om1/internal/inputs"
	"github.com/openmind/om1/internal/logger"
	"github.com/openmind/om1/internal/providers"
)

const (
	descriptorForLLM  = "Home Assistant Device States"
	defaultPollPeriod = 1 * time.Second
	defaultInterval   = 30.0
	fetchTimeout      = 10 * time.Second
)

func init() {
	inputs.Register("HomeAssistantInput", NewHomeAssistantInput)
}

// Config is the decoded plugin configuration for the Home Assistant state input.
type Config struct {
	// BaseURL is the root URL of the Home Assistant instance.
	BaseURL string `json:"base_url"`

	// Token is the Home Assistant long-lived access token.
	Token string `json:"token"`

	// EntityIDs is a comma-separated list of entity IDs to monitor.
	EntityIDs string `json:"entity_ids"`

	// PollInterval is the number of seconds between state polls. 0 means
	// use the default (30s).
	PollInterval float64 `json:"poll_interval"`
}

// StateSensor polls Home Assistant entity states and reports changes.
type StateSensor struct {
	log *zap.Logger

	baseURL      string
	token        string
	entityIDs    []string
	pollInterval time.Duration

	mu           sync.Mutex
	messages     []inputs.Message
	lastPollTime time.Time
	lastStates   map[string]string
	stopped      bool
}

// NewHomeAssistantInput constructs a StateSensor from a decoded config map.
func NewHomeAssistantInput(configMap map[string]any) (inputs.Sensor, error) {
	var cfg Config
	if b, err := json.Marshal(configMap); err == nil {
		_ = json.Unmarshal(b, &cfg)
	}

	log := logger.Get().Named("HomeAssistantInput")

	if cfg.BaseURL == "" {
		log.Warn("base_url not provided")
	}
	if cfg.Token == "" {
		log.Warn("token not provided")
	}

	var entityIDs []string
	for _, e := range strings.Split(cfg.EntityIDs, ",") {
		e = strings.TrimSpace(e)
		if e != "" {
			entityIDs = append(entityIDs, e)
		}
	}
	if len(entityIDs) == 0 {
		log.Warn("no entity_ids configured")
	}

	interval := defaultInterval
	if cfg.PollInterval > 0 {
		interval = cfg.PollInterval
	}

	return &StateSensor{
		log:          log,
		baseURL:      strings.TrimRight(cfg.BaseURL, "/"),
		token:        cfg.Token,
		entityIDs:    entityIDs,
		pollInterval: time.Duration(interval * float64(time.Second)),
		lastStates:   map[string]string{},
	}, nil
}

// fetchState fetches the current state of a single entity from Home Assistant.
func (s *StateSensor) fetchState(ctx context.Context, entityID string) (map[string]any, error) {
	if s.baseURL == "" || s.token == "" {
		return nil, nil
	}

	url := fmt.Sprintf("%s/api/states/%s", s.baseURL, entityID)

	reqCtx, cancel := context.WithTimeout(ctx, fetchTimeout)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+s.token)
	req.Header.Set("Content-Type", "application/json")

	resp, err := httpclient.Default().Do(req)
	if err != nil {
		s.log.Error("network error fetching entity", zap.String("entity_id", entityID), zap.Error(err))
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		s.log.Error("error fetching entity",
			zap.String("entity_id", entityID),
			zap.Int("status", resp.StatusCode),
		)
		return nil, fmt.Errorf("home_assistant input: status %d", resp.StatusCode)
	}

	var state map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&state); err != nil {
		return nil, err
	}
	return state, nil
}

// Poll fetches states for all configured entities, but only once per
// pollInterval; it returns nil in between, matching the throttled polling
// behavior of the original implementation.
func (s *StateSensor) Poll(ctx context.Context) (any, error) {
	s.mu.Lock()
	elapsed := time.Since(s.lastPollTime)
	shouldPoll := elapsed >= s.pollInterval
	if shouldPoll {
		s.lastPollTime = time.Now()
	}
	s.mu.Unlock()

	if !shouldPoll {
		return nil, nil
	}

	if len(s.entityIDs) == 0 {
		return nil, nil
	}

	var states []map[string]any
	for _, entityID := range s.entityIDs {
		state, err := s.fetchState(ctx, entityID)
		if err != nil || state == nil {
			continue
		}
		states = append(states, state)
	}

	if len(states) == 0 {
		return nil, nil
	}
	return states, nil
}

// Listen starts a goroutine that polls at defaultPollPeriod and forwards
// non-nil results to the output channel.
func (s *StateSensor) Listen(ctx context.Context) (<-chan any, error) {
	out := make(chan any)
	go func() {
		defer close(out)
		defer s.Stop()

		ticker := time.NewTicker(defaultPollPeriod)
		defer ticker.Stop()

		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
			}

			raw, err := s.Poll(ctx)
			if err != nil || raw == nil {
				continue
			}

			select {
			case out <- raw:
			case <-ctx.Done():
				return
			}
		}
	}()
	return out, nil
}

// formatState formats a single entity state into human-readable text.
func formatState(state map[string]any) string {
	entityID, _ := state["entity_id"].(string)
	if entityID == "" {
		entityID = "unknown"
	}
	currentState, _ := state["state"].(string)
	if currentState == "" {
		currentState = "unknown"
	}
	attributes, _ := state["attributes"].(map[string]any)
	if attributes == nil {
		attributes = map[string]any{}
	}
	friendlyName, _ := attributes["friendly_name"].(string)
	if friendlyName == "" {
		friendlyName = entityID
	}

	parts := []string{fmt.Sprintf("%s (%s) is %s", friendlyName, entityID, currentState)}

	if brightness, ok := attributes["brightness"].(float64); ok {
		pct := int(brightness/255*100 + 0.5)
		parts = append(parts, fmt.Sprintf("brightness %d%%", pct))
	}

	if colorName, ok := attributes["color_name"].(string); ok {
		parts = append(parts, fmt.Sprintf("color %s", colorName))
	}

	if temperature, ok := attributes["temperature"]; ok {
		parts = append(parts, fmt.Sprintf("temperature %v°C", temperature))
	}

	if currentTemp, ok := attributes["current_temperature"]; ok {
		parts = append(parts, fmt.Sprintf("current temperature %v°C", currentTemp))
	}

	return strings.Join(parts, ", ")
}

// RawToText compares the incoming states against the last known states and
// builds a Message describing only the entities that changed.
func (s *StateSensor) RawToText(_ context.Context, raw any) (*inputs.Message, error) {
	states, ok := raw.([]map[string]any)
	if !ok || len(states) == 0 {
		return nil, nil
	}

	s.mu.Lock()
	var changed []map[string]any
	for _, state := range states {
		entityID, _ := state["entity_id"].(string)
		currentState, _ := state["state"].(string)
		if s.lastStates[entityID] != currentState {
			changed = append(changed, state)
			s.lastStates[entityID] = currentState
		}
	}
	s.mu.Unlock()

	if len(changed) == 0 {
		return nil, nil
	}

	lines := make([]string, 0, len(changed))
	for _, state := range changed {
		lines = append(lines, formatState(state))
	}
	text := "Smart home device updates: " + strings.Join(lines, "; ")

	msg := inputs.NewMessage(text)

	s.mu.Lock()
	s.messages = append(s.messages, *msg)
	s.mu.Unlock()

	return msg, nil
}

// FormattedLatestBuffer formats and clears the latest buffered message.
func (s *StateSensor) FormattedLatestBuffer() string {
	s.mu.Lock()
	defer s.mu.Unlock()

	if len(s.messages) == 0 {
		return ""
	}

	latest := s.messages[len(s.messages)-1]
	result := fmt.Sprintf("\nINPUT: %s\n// START\n%s\n// END\n", descriptorForLLM, latest.Message)

	ts := time.Unix(0, int64(latest.Timestamp*1e9))
	providers.IO().AddInput(descriptorForLLM, latest.Message, ts)
	s.messages = nil

	return result
}

// Stop marks the sensor as stopped. Safe to call multiple times.
func (s *StateSensor) Stop() {
	s.mu.Lock()
	if s.stopped {
		s.mu.Unlock()
		return
	}
	s.stopped = true
	s.mu.Unlock()

	s.log.Info("stopping sensor")
}
